Update file types supported by `drive_download()`

ateucher opened the pull request:
Since Google Docs now [supports exporting to Markdown](https://workspaceupdates.googleblog.com/2024/07/import-and-export-markdown-in-google-docs.html), it seemed like a good idea to enable that support in `drive_download()`.

The only code I changed manually was to add the default `.md` extension for `text/markdown` mimetype in `data-raw/extension-mime-type-defaults.csv`. Then I just ran `data-raw/mime-types.R` to pull in the Google MIME types and create the csv files in `inst/extdata/data/`.

This does work now:

``` r
library(googledrive)

drive_auth(email = "andy@andyteucher.ca")

drive_download(
  "https://docs.google.com/document/d/1EK6WZyEUfdy6sj3GjLII20Ok7qTcmRD-vv-osfHnkPU",
  path = "my_document.md",
)
#> File downloaded:
#> • 'Test for google drive' <id: 1EK6WZyEUfdy6sj3GjLII20Ok7qTcmRD-vv-osfHnkPU>
#> Saved locally as:
#> • 'my_document.md'
```

<sup>Created on 2025-07-09 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

ateucher committed:
Add default extension for markdown files

ateucher committed:
Run data-raw/mime-types.R

ateucher commented:
Converted to draft - I'm going to do a bit more work around extensions and defaults

ateucher committed:
Add failing test for #465

* If mimetype does not have an extension in mime_tbl.csv NA is appended to the filename

ateucher committed:
Don't add extension if none in mime_types.csv

ateucher commented:
I don't think there is any more to do around adding default extensions to mimetypes. Most of the changes here come from changes in `mime::mimemap` and it seems sensible to just accept them.

I did however discover that if a mimetype does not have an extension in `inst/extdata/data/mime_tbl.csv`, `.NA` is appended to the given filename if that mimetype is specified in `type`. Eg., 

``` r
drive_download(
  "https://docs.google.com/document/d/1EK6WZyEUfdy6sj3GjLII20Ok7qTcmRD-vv-osfHnkPU",
  type = "text/x-markdown",
  path = "my_document.md",
  overwrite = TRUE
)
#> File downloaded:
#> • 'Test for google drive' <id: 1EK6WZyEUfdy6sj3GjLII20Ok7qTcmRD-vv-osfHnkPU>
#> Saved locally as:
#> • 'my_document.md.NA'
```

I added a fix and a test for this, so now it does not add `.NA`:

``` r
drive_download(
  "https://docs.google.com/document/d/1EK6WZyEUfdy6sj3GjLII20Ok7qTcmRD-vv-osfHnkPU",
  type = "text/x-markdown",
  path = "my_document.md",
  overwrite = TRUE
)
#> File downloaded:
#> • 'Test for google drive' <id: 1EK6WZyEUfdy6sj3GjLII20Ok7qTcmRD-vv-osfHnkPU>
#> Saved locally as:
#> • 'my_document.md'
```

<sup>Created on 2025-07-10 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

jennybc committed:
One script per csv file

jennybc committed:
Revisit the prep of the table for MIME types <--> extensions

jennybc commented on inst/extdata/data/translate_mime_types.csv:
```diff
@@ -45,9 +49,11 @@ import,application/vnd.google-apps.document,image/png,NA
 import,application/vnd.google-apps.document,text/plain,NA
 import,application/vnd.google-apps.document,text/richtext,NA
 import,application/vnd.google-apps.document,text/rtf,NA
+import,application/vnd.google-apps.document,text/x-markdown,NA
```
`text/x-markdown` appears to be a rather weird/legacy MIME type, but if it's one of the official ones (and it is), I agree we need to plan for it. This is a note to myself.

jennybc committed:
Update snapshot

jennybc committed:
Test the utility instead

The test won't work as written anymore, using the "text/x-markdown" MIME type. Instead of modifying it, I tested the filepath utility itself.

jennybc committed:
Make Air happy

jennybc committed:
Update data-raw/mime-types-google.R

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

jennybc committed:
Merge branch 'main' into ateucher-update-download-mimetypes

jennybc committed:
Add a new example markdown file

jennybc committed:
Add back a test of drive_download()

jennybc committed:
Add a test for upload of md to doc

jennybc committed:
Add NEWS bullet

ateucher commented:
Great, thanks so much for finishing it off!

jrosell commented:
@ateucher  thanks for your PR!.

It would have been awesome if the code were not lost in the process of upload + download, but I guess it's a google issue and not an implementation issue.

``` r
pak::pak("tidyverse/googledrive")
#> ℹ Loading metadata database
#> ✔ Loading metadata database ... done
#> 
#> 
#> ✔ All system requirements are already installed.
#> 
#> ℹ No downloads are needed
#> ✔ 1 pkg + 24 deps: kept 24 [4.8s]
```

```` r
library(googledrive)

tfile <- tempfile(fileext = ".md")
code <- r"(

# Title

```r
1+1
```

text
)"

writeLines(code, tfile, sep = "\n")
cat(paste0(readLines(tfile), "\n"))
#> 
#>  
#>  # Title
#>  
#>  ```r
#>  1+1
#>  ```
#>  
#>  text
#> 
````

``` r

options(gargle_oauth_email = Sys.getenv("GOOGLE_EMAIL"))
drive_auth_configure(
  path = Sys.getenv("GOOGLE_API_CREDENTIALS_OAUTH20_CLIENT_ID_DESTOP_FILE")
)
drive_upload(tfile, type = "document")
#> Local file:
#> • '/tmp/Rtmp2Lg9bf/file27bff5042c5a8.md'
#> Uploaded into Drive file:
#> • 'file27bff5042c5a8.md' <id: 1aX6tY9RIyMiq_RB1LSDeai7judGcVhGNKi4JhEdD7f8>
#> With MIME type:
#> • 'application/vnd.google-apps.document'


drive_download(
  basename(tfile),
  path = "markdown-googledrive.md",
  overwrite = TRUE
)
#> File downloaded:
#> • 'file27bff5042c5a8.md' <id: 1aX6tY9RIyMiq_RB1LSDeai7judGcVhGNKi4JhEdD7f8>
#> Saved locally as:
#> • 'markdown-googledrive.md'
cat(paste0(readLines("markdown-googledrive.md"), "\n"))
#> # Title
#>  
#>  1+1
#>  
#>  text
```

<sup>Created on 2025-09-10 with [reprex v2.1.1.9000](https://reprex.tidyverse.org)</sup>

jennybc commented:
If I send this markdown:

````
# Title

```r
1+1
```

```rust
fn main() { println!("Hello, world!"); }
```

text
````

then re-download it, I get:

````
# Title

```
1+1
```

```rust
fn main() { println!("Hello, world!"); }
```

text  
````

The code fences are retained. But the language identifier for `r` is lost, where the one for `rust` is retained. I think this is a Google problem, because R is not available in this dropdown list of languages:


https://github.com/user-attachments/assets/a7e909c7-8993-401c-ac44-a83ac45eec6c

It would be great to include R there, since it is more popular than some of these languages. But it's beyond the scope of what googledrive can do.
